Add bad pixel monitor

bhilbert4 opened the pull request:
This PR adds the bad pixel monitor to JWQL.

The bad pixel monitor has two distinct components:

- Bad pixels identified using internal flat field files
- Bad pixels identified using dark current files

The `spacetelescope:jwst_reffiles` package is used for both. This package was put together by a working group composed of representatives from all instruments, which came up with a common algorithm for all instruments.

The bad pixel monitor is based on the dark current monitor, in that MAST is queried and when enough new files are present, the data are copied to a working directory. The appropriate calibration pipeline steps are run, and then the bad pixel search is performed. The resulting bad pixels are compared to those in the current bad pixel reference file in CRDS. Lists of newly identified bad pixels (organized by type of bad pixel) are saved to the appropriate database tables.

pep8speaks commented:
Hello @bhilbert4, Thank you for submitting the Pull Request !



* In the file [`jwql/instrument_monitors/common_monitors/dark_monitor.py`](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py):

> [Line 71:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L71): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (112 > 100 characters)
> [Line 72:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L72): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (112 > 100 characters)
> [Line 73:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L73): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (106 > 100 characters)
> [Line 74:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L74): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (115 > 100 characters)
> [Line 75:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L75): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 100 characters)
> [Line 79:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L79): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (106 > 100 characters)
> [Line 142:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L142): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 100 characters)
> [Line 354:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L354): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 100 characters)
> [Line 358:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L358): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (106 > 100 characters)
> [Line 369:9](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L369): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 369:9](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L369): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 369:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L369): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (110 > 100 characters)
> [Line 370:9](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L370): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 370:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L370): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (112 > 100 characters)
> [Line 371:9](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L371): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 371:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L371): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (113 > 100 characters)
> [Line 372:9](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L372): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 372:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L372): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (117 > 100 characters)
> [Line 373:9](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L373): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 377:9](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L377): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 378:9](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L378): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 381:5](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L381): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 381:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L381): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 100 characters)
> [Line 494:46](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L494): [E712](https://duckduckgo.com/?q=pep8%20E712) comparison to True should be 'if cond is True:' or 'if cond:'
> [Line 596:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L596): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (120 > 100 characters)
> [Line 610:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L610): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (116 > 100 characters)
> [Line 629:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L629): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (101 > 100 characters)
> [Line 644:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L644): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (101 > 100 characters)
> [Line 645:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L645): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (124 > 100 characters)
> [Line 667:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L667): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (121 > 100 characters)
> [Line 668:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L668): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (123 > 100 characters)
> [Line 682:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L682): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (128 > 100 characters)
> [Line 692:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L692): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (106 > 100 characters)
> [Line 795:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L795): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (102 > 100 characters)
> [Line 797:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L797): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 100 characters)
> [Line 802:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L802): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (104 > 100 characters)
> [Line 811:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L811): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (113 > 100 characters)
> [Line 834:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L834): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (107 > 100 characters)
> [Line 876:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L876): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (105 > 100 characters)
> [Line 1025:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L1025): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (102 > 100 characters)
> [Line 1032:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L1032): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (110 > 100 characters)
> [Line 1033:42](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L1033): [W503](https://duckduckgo.com/?q=pep8%20W503) line break before binary operator
> [Line 1041:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L1041): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (129 > 100 characters)
> [Line 1042:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L1042): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (127 > 100 characters)
> [Line 1043:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L1043): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (102 > 100 characters)
> [Line 1046:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L1046): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (147 > 100 characters)
> [Line 1057:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L1057): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 100 characters)
> [Line 1058:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L1058): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (107 > 100 characters)
> [Line 1059:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L1059): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (108 > 100 characters)
> [Line 1064:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/instrument_monitors/common_monitors/dark_monitor.py#L1064): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (101 > 100 characters)



* In the file [`jwql/utils/constants.py`](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/utils/constants.py):

> [Line 52:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/utils/constants.py#L52): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (102 > 100 characters)
> [Line 53:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/utils/constants.py#L53): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (109 > 100 characters)
> [Line 82:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/utils/constants.py#L82): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (105 > 100 characters)
> [Line 169:21](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/utils/constants.py#L169): [E126](https://duckduckgo.com/?q=pep8%20E126) continuation line over-indented for hanging indent

* In the file [`jwql/utils/crds_tools.py`](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/utils/crds_tools.py):

> [Line 170:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/utils/crds_tools.py#L170): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (132 > 100 characters)
> [Line 179:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/utils/crds_tools.py#L179): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (132 > 100 characters)
> [Line 190:27](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/utils/crds_tools.py#L190): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

* In the file [`jwql/utils/mast_utils.py`](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/utils/mast_utils.py):

> [Line 103:101](https://github.com/spacetelescope/jwql/blob/761e3920e9f5c217cb81c75eefa12d507c9ec43c/jwql/utils/mast_utils.py#L103): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 100 characters)

If you have not done so, please consult the [`jwql` Style Guide](https://github.com/spacetelescope/jwql/blob/master/style_guide/README.md)

bhilbert4 commented:
The initial commit contains most (I think) of the necessary code, but nothing has yet been tested at all.

bhilbert4 commented:
Testing (restricted to NIRCam NRCA1_FULL) shows this now running to completion successfully.

bourque commented on jwql/instrument_monitors/common_monitors/bad_pixel_monitor.py:
```diff
@@ -0,0 +1,1066 @@
+            for uncal_file, rate_file in zip(illuminated_raw_files, illuminated_slope_files):
+                self.get_metadata(uncal_file)
+                if rate_file == 'None':
+                    jump_output, rate_output, junk = pipeline_tools.calwebb_detector1_save_jump(uncal_file, self.data_dir,
```
I often see people use `_` for variable names that are not going to be used, it might be a standard way of dealing with that.

bhilbert4 replied:
Done

bourque commented on jwql/instrument_monitors/common_monitors/bad_pixel_monitor.py:
```diff
@@ -0,0 +1,1066 @@
+                    jump_output, rate_output, junk = pipeline_tools.calwebb_detector1_save_jump(uncal_file, self.data_dir,
+                                                                                          ramp_fit=True, save_fitopt=False)
+                    if self.nints > 1:
+                        #illuminated_slope_files[index] = rate_output.replace('rate', 'rateints')
```
This can be removed if it is not needed

bhilbert4 replied:
Removed

bourque commented on jwql/instrument_monitors/common_monitors/bad_pixel_monitor.py:
```diff
@@ -0,0 +1,1066 @@
+        # in order to make things easier to read.
+        output_file = '{}_{}_{}_{}_bpm.fits'.format(self.instrument, self.aperture, self.dark_query_start, self.query_end)
+        output_file = os.path.join(self.output_dir, output_file)
+        bad_pixel_mask.bad_pixels(flat_slope_files=illuminated_slope_files, dead_search_type=dead_search_type,
```
Holy parameter space!

bhilbert4 replied:
This is what happens when you get reps from all the instruments in a room every week for months to discuss the *best* way to find bad pixels.....You end up with 17 *best* ways.

bourque commented on jwql/instrument_monitors/common_monitors/bad_pixel_monitor.py:
```diff
@@ -0,0 +1,1066 @@
+                                  history='This file was created by JWQL', quality_check=False)
+
+        # Read in the newly-created bad pixel file
+        set_permissions(output_file)
```
Thanks for remembering to do that!

bourque commented on jwql/utils/crds_tools.py:
```diff
@@ -0,0 +1,190 @@
+    """
+    # IMPORTANT: Import of crds package must be done AFTER the environment
+    # variables are set in the functions above
+    import crds
```
I think this is a new dependency for the project.  Could you add the appropriate lines to `setup.py`, `requirements.txt`, and `environment-3.6.yml`?

bhilbert4 replied:
Done

bourque commented on jwql/utils/crds_tools.py:
```diff
@@ -0,0 +1,190 @@
+    """Create a dictionary to be used as input to the CRDS getreferences
+    function from the nested dictionary created when a standard Mirage
+    input yaml file is read in.
+    Parameters
```
Missing a blank line here

bhilbert4 replied:
Inserted.

bourque commented on jwql/instrument_monitors/common_monitors/bad_pixel_monitor.py:
```diff
@@ -0,0 +1,1066 @@
+from astropy.io import ascii, fits
+from astropy.time import Time
+from jwst.datamodels import dqflags
+from jwst_reffiles.bad_pixel_mask import bad_pixel_mask
```
I think this is a new dependency for the project.  Could you add it appropriately to `setup.py`, `requirements.txt`, and `environment-3.6.yml`?

bhilbert4 replied:
Done

bhilbert4 commented on jwql/instrument_monitors/common_monitors/bad_pixel_monitor.py:
```diff
@@ -0,0 +1,1066 @@
+
+
+
+Development notes:
```
Updated and removed statement that is no longer true.

bhilbert4 commented on jwql/instrument_monitors/common_monitors/bad_pixel_monitor.py:
```diff
@@ -0,0 +1,1066 @@
+            # Identify which database tables to use
+            self.identify_tables()
+
+            # Get a list of all possible apertures from pysiaf
```
Created get_possible_apertures() and moved the aperture code in there.

bhilbert4 commented on jwql/utils/mast_utils.py:
```diff
@@ -0,0 +1,291 @@
+from jwql.utils.logging_functions import configure_logging, get_log_status
+
+
+def initialize_instrument_monitor(module):
```
Weird. I feel like I did that because of circular import issues, although looking at the code now, I don't see any circular imports. I've removed the copies of the functions from mast_utils.py

bhilbert4 commented on jwql/utils/mast_utils.py:
```diff
@@ -0,0 +1,291 @@
+    return start_time, log_file
+
+
+def mast_query_orig(instrument, aperture, templates, start_date, end_date):
```
Yes. mast_query() is more flexible.

bhilbert4 commented on jwql/instrument_monitors/common_monitors/bad_pixel_monitor.py:
```diff
@@ -0,0 +1,1066 @@
+                print('DARKS')
+                for entry in new_dark_entries:
+                    print(entry['detector'], entry['apername'], entry['filter'], entry['readpatt'])
+                #print('FLATS - PREV')
```
Removed all the print statements left over from development.

bhilbert4 commented on jwql/utils/mast_utils.py:
```diff
@@ -0,0 +1,291 @@
+        if grating is not None:
+            parameters["grating"] = grating
+        if readpattern is not None:
+            parameters["readpatt"] = readpattern
```
Tell me about it. So many options...

bhilbert4 commented on jwql/instrument_monitors/common_monitors/bad_pixel_monitor.py:
```diff
@@ -0,0 +1,1066 @@
+                if len(flat_uncal_files) > 0:
+                    flat_uncal_files = sorted(list(set(flat_uncal_files)))
+
+                if len(flat_uncal_files) < flat_file_count_threshold:
```
If that were made an `elif`, then if enough files were found to run the monitor, they would go into the `if` block to be sorted and have repeats removed, but then they would not go into the `else` block, which is what calls the monitor itself. The first `if` statement needs to be separate, so that if any files are found, any repeats will be removed, and then the second `if/else` block can decide if there are enough unique files to call the monitor or not.

Yes, I was assuming that the threshold would always be more than 0, since having a threshold of 0 doesn't really make sense.

bhilbert4 commented on jwql/instrument_monitors/common_monitors/bad_pixel_monitor.py:
```diff
@@ -0,0 +1,1066 @@
+                flat_uncal_files = locate_uncal_files(new_flat_entries)
+                dark_uncal_files = locate_uncal_files(new_dark_entries)
+
+                # Remove duplicates and check if there are enough new flat
```
Done.

bhilbert4 commented on jwql/instrument_monitors/common_monitors/bad_pixel_monitor.py:
```diff
@@ -0,0 +1,1066 @@
+#! /usr/bin/env python
+
+"""This module contains code for the bad/dead pixel monitor, XXX
+XXXXXXXXXXXXXX THIS DOCSTRING NEEDS TO BE UPDATED
```
Docstrings updated.

bhilbert4 commented:
@bourque I think I've addressed all of your comments. Note that in `setup.py`,  and the environment file I bumped up the version of `jwst` to 0.16.2.
